src/graph/query_graph.c

```diff
@@ -93,6 +102,99 @@ static void _BuildQueryGraphAddEdge(QueryGraph *qg, const cypher_astnode_t *ast_
 	else QueryGraph_ConnectNodes(qg, dest, src, edge);
 }
 
+// Extracts node from 'qg' and places a copy of into 'graph'
+static void _QueryGraph_ExtractNode(const QueryGraph *qg, QueryGraph *graph,
+		AST *ast, const cypher_astnode_t *ast_node) {
+
+	// Validate inputs.
+	ASSERT(qg != NULL && graph != NULL && ast != NULL && ast_node != NULL);
+
+	// See if node is already in 'graph'.
+	const char *alias = AST_GetEntityName(ast, ast_node);
+	QGNode *n = QueryGraph_GetNodeByAlias(graph, alias);
+
+	if(n == NULL) {
+		// Node is missing from 'graph', try getting it from 'qg'.
+		n = QueryGraph_GetNodeByAlias(qg, alias);
+		if(n == NULL) {
+			// Node is missing from 'qg', create it.
+			_QueryGraphAddNode(graph, ast_node);
+		} else {
+			// Add a clone of the original node.
+			n = QGNode_Clone(n);
+
+			// Clear node label information.
+			n->label = NULL;
+			n->labelID = GRAPH_NO_LABEL;
+
+			QueryGraph_AddNode(graph, n);
+			// Set node label information.
+			_QueryGraphSetNodeLabel(n, ast_node);
+		}
+	}
+}
+
+// Extracts edge from 'qg' and places a copy of into 'graph'
+static void _QueryGraph_ExtractEdge(const QueryGraph *qg, QueryGraph *graph,
+		QGNode *left, QGNode *right, AST *ast, const cypher_astnode_t *ast_edge) {
+	const char *alias = AST_GetEntityName(ast, ast_edge);
+
+	// Validate input, edge shouldn't be in graph.
+	ASSERT(left != NULL && right != NULL);
+	ASSERT(QueryGraph_GetEdgeByAlias(graph, alias) == NULL);
+
+	// Try getting edge from 'qg'
+	QGEdge *e = QueryGraph_GetEdgeByAlias(qg, alias);
+	if(e == NULL) {
+		// Edge is missing from 'qg' create it.
+		_QueryGraphAddEdge(graph, ast_edge, left, right);
+	} else {
+		// Edge is part of 'qg', add a clone of it to 'graph'
+		QGNode *src = QueryGraph_GetNodeByAlias(graph, e->src->alias);
+		QGNode *dest = QueryGraph_GetNodeByAlias(graph, e->dest->alias);
+		ASSERT(src != NULL && dest != NULL);
```

Comment:
This assertion fails with the query:
`MATCH (a)-[e]->(b) WITH a WHERE (a)-[e]->(:L1) RETURN a`
Due, I believe, to the back-propagation of path filters to previous scopes described at 119.

Reply:
Great catch!